### ISSUE TITLE
Genie2 - Cannot make final step work... everything leading up to it is confirmed

### ISSUE THREAD

**Issue by Ernster93** [bklavet/Echo-skill-to-control-Directv#72]
Hi...  I have the latest Genie version, the Genie2, with 3 minis.  I use the Custom Slot Skill for Mini Genie and utilize the MAC codes on line 12 for the index as well as enter my correct wan_ip on line 8.  Everything seems to work fine all the way up to the test step.  At the test step, I get the the following result:  The remote endpoint could not be called, or the response it returned was invalid.  I even tested the pause and hold URL you provided for each of my minis and they do freeze at when I click confirm.  My settings are set for allow external devices and I have reviewed everything you have listed under https://github.com/bklavet/Echo-skill-to-control-Directv/issues/11.  I did have to create a new Lambda, but I'm confident I was able to manage through that correctly.  Help!

**Comment by vmweaver** [comment]
Did you have a look at https://github.com/bklavet/Echo-skill-to-control-Directv/issues/12 ?
Please make sure you can do the last part of step 3 with your WAN ip. Did you test the PAUSE functions with your local ip address?
 This will confirm your port forwarding is working fine on your external (internet) address.

**Comment by Ernster93** [comment]
This could be where I am having my problem.  I can use the pause command with the local IP using 8080, but seem to be having trouble when I take my cell phone, walk down the street away from my WiFi and try to use the WAN IP, nothing really happens.  After a couple of minutes, I get timed out.  I notice the last line of step 3 essentially mentions that in this case my ISP is blocking my firewall.  How do I get that fixed?  I found a website, https://itstillworks.com/unblock-port-8629973.html, which tells me how to unblock a port, but when I get to the "Allow apps to communicate through Windows Defender Firewall", a number of "Allowed apps and features" are listed.  I'm not sure which one to pick.  I don't have DirecTV listed there.  Am I doing the right thing?

**Comment by bklavet** [comment]
Hello Ernster,  I don't think this would be a windows defender thing.  There is a router/ modem that needs to have port 8080 forwarded.  Check your modem/ router settings and  see if you have any luck.  You are correct this looks like a port forwarding problem.

**Comment by vmweaver** [comment]
Agreed. Who is your service provider and what model of router do you have?  As an FYI, all you need to do is disable WIFI on you phone, unless you need the exercise. ;-)

**Comment by Ernster93** [comment]
Optimum is the service provider and Arris TM822 is the modem.  I can log onto my stuff at router.optimum.net webpage and I find Port Forwarding under the Advanced Settings, but it only seems to allow me to set up Port Forwarding with the Local Host and not the WAN ID.  Might I have to call Optimum to see if they can help me?  I just don't understand how to either "unblock" an ISP or navigate through the Port Forwarding site to allow me to utilize the WAN ID.

**Comment by vmweaver** [comment]
Can you possibly attach a screen shot of the port forwarding user interface?

**Comment by Ernster93** [comment]
![port management](https://user-images.githubusercontent.com/32876160/35493931-f41e93d2-0484-11e8-9918-e8748c4e402f.PNG)

**Comment by Ernster93** [comment]
![port forwarding1](https://user-images.githubusercontent.com/32876160/35493936-01a62538-0485-11e8-99a5-225f4532dc04.PNG)

**Comment by Ernster93** [comment]
![port forwarding2](https://user-images.githubusercontent.com/32876160/35493937-075c6546-0485-11e8-9f03-56aeb241dd45.PNG)

**Comment by Ernster93** [comment]
![port triggering1](https://user-images.githubusercontent.com/32876160/35493942-0c1eee64-0485-11e8-9482-ffafb45546e8.PNG)
![port triggering2](https://user-images.githubusercontent.com/32876160/35493944-0e76e90a-0485-11e8-8153-c007dc564b7b.PNG)

**Comment by Ernster93** [comment]
I believe the above is what you are asking for.  Anything labeled DirecTV I added on my own.  I think I gave you the wrong screenshot though for Port Triggering... I didn't give you what I entered for DirecTV... so here it is...
![port triggering3](https://user-images.githubusercontent.com/32876160/35494004-6ff21042-0485-11e8-961f-e799f8379284.PNG)

**Comment by vmweaver** [comment]
Thanks.  Looks good, but you shouldn't need port triggering at all, so I would remover that configuration. 
Is there any place in the router configuration that shows what your internet port ip address is and how does that compare to the address that Google gives you if you search for "what's my ip"?

**Comment by Ernster93** [comment]
When I click on a link for WAN Setup, it lists "Internet IP Address" as 69.122.57.141 which is the same thing I get in Google when I type in What's My IP.  I don't know if this means anything but I also have a Default Gateway IP under the WAN Setup of 69.122.48.1 and there is two lines for DNS Servers... 65.19.96.252 and 65.19.96.253.... by the way... thanks for helping me.

**Comment by vmweaver** [comment]
Did you remove the port triggering config? And did you put the DTV address in the dmz config? If so, please remove that too. Just need the port forwarding rule in place.

Also to confirm that going to http://192.168.1.18:8080/tv/getTuned returns info about the current show?

**Comment by Ernster93** [comment]
yes

**Comment by vmweaver** [comment]
Okay, so you just have the port forwarding set up? No triggering or dmz?
 Do you have a screen to configure firewall rules also? Some routers do and some dont.
If so, can you attach a screenshot of that?

**Comment by Ernster93** [comment]
Ya know... that question about http://192.168.1.18:8080/tv/getTuned... nothing at all comes up when I load that (I get Problem Loading Page message)... I have found, for whatever reason that using 192.1681.21 actually is what gives me access... for instance ... if I use http://192.168.1.21:8080/info/getLocations... I get the following for all 3 TVs I have (assuming they are all on which currently they are):
![tv locations](https://user-images.githubusercontent.com/32876160/35495208-cd363082-048d-11e8-9f1f-f6f3d552ded3.PNG)

**Comment by Ernster93** [comment]
http://192.168.1.21:8080/tv/getTuned gives me the following:
![get tuned](https://user-images.githubusercontent.com/32876160/35495258-0b64bfe0-048e-11e8-8351-e50c77fec3d5.PNG)

**Comment by bklavet** [comment]
Port 80 & 8080 Blocking
Because Port 80 and 8080 are often used by malicious software, including viruses and worms, Optimum, like many ISPs (Internet Service Providers), blocks this port for all standard Optimum customers.

Activating Port 80 & 8080
If you subscribe to Business Optimum 50*/60 service and above and would like to utilize Port 80 and/or 8080, you'll need to activate them. Please note that you will need to be on a computer that is connected through the same modem associated with your Optimum Business service account. Simply follow these steps:

Go to https://www.optimum.com/business.
Sign in with your Optimum ID and password
Hover your mouse over Internet and click Network Configuration
Under Port Configuration, click on Settings
Under Port 80 or 8080, click On to enable
Click Submit
Unplug your modem's power and plug it back in. You will need to reboot your modem for the changes to take effect.

Note: If you swap your modem, you may need to re-enable Ports 80 and 8080 for your new modem.

When using Port 80 and/or 8080 to host your own website, please be diligent by keeping your web server up to date with the latest security patches.

**Comment by vmweaver** [comment]
Ah, so .18 doesn't seem to be your DTV receiver(?). Change your port forwarder to use 192.168.1.21 and see if that works.

**Comment by bklavet** [comment]
my guess is that optimum is blocking port 8080, so you could talk to them and see if the will unblock, or get a pi and go with mawrew19's security option.

**Comment by vmweaver** [comment]
Good catch. 
You should be able to change your port forwarding rule to use port 3000 on the service and then uncheck the box for using the same internal port and change that to be 8080.

**Comment by bklavet** [comment]
http://192.168.1.21:8080/tv/getTuned 

the receiver needs to be turned on to a channel

**Comment by Ernster93** [comment]
I have all 3 TVs turned on at the moment... although as I said, when I use the getTuned web address you mention above, I get Internal Server Error, etc.

**Comment by vmweaver** [comment]
No worries, this is likely just because you have all remotes.
I would stick with /info/getLocations to test.  I think bklavett is right, though,  on your isp blocking 8080. Can u try the 3000 port I mentioned?

**Comment by Ernster93** [comment]
mawrew19... I just did as you suggested and saved the changes.  Now what do I do?
![3000 to 8080](https://user-images.githubusercontent.com/32876160/35495831-ee9454cc-0490-11e8-93ea-66b28d38b1b5.PNG)

**Comment by vmweaver** [comment]
Can you change your host to be 192.168.1.21 on the port forward? Is that the lan ip address of dtv? What device has .21 versus .18?

**Comment by Ernster93** [comment]
With the new Genie2 setup, there is a main server (see https://support.directv.com/equipment/directv-genie-2-dvr-hs17dtv-genie-server if you are not familiar) which by itself is not connected to any TV.  You can then add as many mini's as you need (up to whatever the max is).  My main server, called The Tower by DirecTV, has 192.168.1.21 with MAC B42AOEEFBCD6.  My mini's are as follows (for what it's worth):
2832C5EDDD6A – Living Room  192.168.1.18
A0722C8B2656 – Kitchen   192.168.1.19
8E579BA55264 – Bedroom  192.168.1.6

As you've seen above as I mentioned in my first post, when I want to give the command to pause a screen in my living room for instance, I have used the following format with a Local IP:
http://192.168.1.21:8080/remote/processKey?key=pause&hold=keyPress&clientAddr=2832C5EDDD6A

**Comment by Ernster93** [comment]
I just can't get it to work using the WAN IP.

**Comment by vmweaver** [comment]
Interesting. ..I had not seen this setup before, but it makes sense.
Is your port forwarder currently set to the .21 address? If not, please set that.  


Ah... I see it now.
Go to http://69.122.xxx.yyy:3000/info/getLocations
Replace xxx.yyy with the last 2 octets of your actual external address.

In your index file for the lambda service, replace 8080 with 3000 for the port.

**Comment by Ernster93** [comment]
I updated the Port Forwarder now to the HS17 using the .21 address...  trying to pause using the WAN ID on my phone (using data) so I'm not on my WiFi at all... still no luck.

**Comment by vmweaver** [comment]
I am going to try pause on you tv. Ok?

**Comment by Ernster93** [comment]
OK... let me know when you do...  I assume you are going to try the Living Room TV.

**Comment by vmweaver** [comment]
Yep.. 1 sec

**Comment by Ernster93** [comment]
You did.

**Comment by vmweaver** [comment]
Paused?

**Comment by Ernster93** [comment]
Did you use 69.122.57.141?

**Comment by vmweaver** [comment]
http://69.122.xxx.yyy:3000/remote/processKey?key=pause&hold=keyPress&clientAddr=2832C5EDDD6A

**Comment by Ernster93** [comment]
Ah... 3000.. OK

**Comment by vmweaver** [comment]
Now we should remove the a dresses from comments. Lol

**Comment by Ernster93** [comment]
So... I should be good to go now?

**Comment by Ernster93** [comment]
I see it works on my phone too.

**Comment by vmweaver** [comment]
See my previous comet about index file update to use port 3000.

Should be once you update lambda.

**Comment by Ernster93** [comment]
Oh geez... now I'm nervous again.

**Comment by Ernster93** [comment]
Remind me... which comment are you referring to about index file update and what do I do about Lambda?

**Comment by Ernster93** [comment]
On the index file, wherever it says 8080, do I just change it to say 3000?

**Comment by vmweaver** [comment]
Ok. So the index file has your WAN ip in it and the appid and the mac address of you genie remote.
There is also a port defibed in it. Change the port from 8080 to be 3000.

Then you will zip this file up with the AlexaSkill file. 
Then login to the amazon developer site for lambda and upload it to the DTV lambda function you setup.

May be easier to walk though bklavett 's tutorial again.

**Comment by vmweaver** [comment]
Yep.. change to 3000.

**Comment by Ernster93** [comment]
I will try... made the change to 3000 on the index file.  I will start from scratch, but believe me, I don't need to see the tutorial again... I've seen it 100 times already!  LOL  :-)

**Comment by vmweaver** [comment]
Lol... awesome. Good luck. I am heading to bed.

**Comment by Ernster93** [comment]
Thanks for all your help!  Really and truly I appreciate it!  I hope it works.... will try and let you know...  I want to head to bed very soon too!

**Comment by Ernster93** [comment]
Grr!  I got it to work in the test late last night, but it wasn't working when I asked Alexa at all.  Therefore, I wanted to start over anyway this morning.. and start with my Kitchen TV.  I deleted the "living room" skill from the Developer Console as well as all Lambda Functions related to that skill.  I haven't changed anything in the Port Forwarding (still 3000 to 8080), but I did want to update the Index and the custom slot values to include more stations along with their associated channels.  I made sure that every channel I included on the index was also included on the custom slot values, although I didn't necessarily enter every channel in the exact same order in the index as on the values tab (I didn't think it mattered).  Now, after enter all the new data, I'm once again getting error message when I try to run the test.  Again, I have no problem pausing any of my TVs with the 192.168.1.21 number along with the MAC addresses, but when I try to enter the WAN_IP to pause the screen, it now comes back with a 403 error.  Although I may not be at the desk here for a while waiting for your reply... I meekly request 'help' again if you can.

**Comment by vmweaver** [comment]
I see that your system is still publicly exposed, but doing a
/info/getLocations onlly shows your "server"/Tower.
Can I try pausing one of your TVs (living room)?

  Can you zip up your Index file, your required additions to the channels
and email me? geekweaver (at) gmail.com
Would like to take a look and see what is going on.

**Comment by Ernster93** [comment]
Hi... sorry it's taken awhile... had to be away from the home.  Just sent e-mail as you requested.  At the moment, I have Living Room TV on for you.  I may not necessarily be sitting here every second (and I don't expect you to either).  Thanks for helping at your own pace.

**Comment by mawrew19** [comment]
One interesting thing I noticed as I am digging into this is that the URL call to specify a clientAddr for your receivers is 
"?clientAddr="    and the "standard" call is "&clientAddr="
This seems to be a bit of the OLD syntax for the SHEF API, but not sure.
I sent you a private message about this and what to try.  Let me know how that goes.

**Comment by Ernster93** [comment]
yes... I just replied to that private message... now I'm seeing this one

**Comment by Ernster93** [comment]
you did the pause thing somehow

**Comment by vmweaver** [comment]
Yes... I did..  :-)  Remember that since we have a special port forwarder rule in place you will ALWAYS use port 3000 when using your WAN address and always use port 8080 when using your internal address.

**Comment by Ernster93** [comment]
yep.... just replied via e-mail.  I got it to work.  I think talking via this method is better... but sharing as much private info as possible is better via e-mail, no?

**Comment by Ernster93** [comment]
The thing I got to work was the pause function... just to be clear.

**Comment by Ernster93** [comment]
So... with the change to the var dtv_Mac line in the index file to now read with the "?", should I try and create  a new Alexa Skill and Lambda function to see if I can get everything to test correctly now... or since I added channels and such, do you want to see my custom slot values and index page first?

**Comment by vmweaver** [comment]
Sorry.. in one of my private messages I mentioned to leave it as is.. that I was wrong in my assessment. So very sorry.
Let me review your index and custom slot stuff real quick.  
So.. please leave the "&clientAddr"

**Comment by Ernster93** [comment]
(Removed ZIP file)

**Comment by vmweaver** [comment]
Been working with Ernter93 offline and the primary resolution for getting things to "mostly work" (still have to tweak some custom channels and such) was the discovery that his household had multiple Amazon accounts and I believe we should update our documentation to ensure we are clear that the account linked to the Echo device is the same as the one used for Alexa Skills Kit.  Apparently the Lambda Skill can be related to any account.
@bklavet : May need your expertise in helping @Ernster93 work though utterances and custom channels in the index file.

**Comment by bklavet** [comment]
Cool!  Wow with different profiles, I think you would have to do a developer skill for each account and tie it to the one lambda function arn

Let me know the channels and networks and I can see what I can come up with.

**Comment by Ernster93** [comment]
[custom slot skills.xlsx](https://github.com/bklavet/Echo-skill-to-control-Directv/files/1689369/custom.slot.skills.xlsx)

bklavet... Here is my list of custom slot values.  It's a combination of yours and what I've added.  The ones that I have trouble with (both a combination of yours and mine), I have documented on this sheet.  If you need the index, let me know.  Just wasn't sure if I should attach it here with my WAN IP address and MAC listed.

This is awesome BTW!

**Comment by bklavet** [comment]
alright remove the wan and macs and app id from your index and attach it. to the thread.
thanks

**Comment by Ernster93** [comment]
bklavet... I screwed something up and now I think I have to completely start over.  I tried to fix a couple of things on my own... and now I can't even get the test to work.  I will get back to you as soon as possible.

**Comment by Ernster93** [comment]
bklavet... OK ... got myself working again.  See attached for my index zip file.

**Comment by bklavet** [comment]
ok, so this may resolve some of it,  see if the locals work for you now.
you will want to add your mac, app id, and wan to the attached index.
check out the custom slot values attached.  I have altered it for the index.  

So here's the rub.  We don't have any control on how alexa passes what it passes to the aws function.  we can only plan for what might be sent, and the custom slot values is a good way to get the odds better,  but I have seen first hand what works fine for months suddenly stops because alexa is passing a slightly different value.  

Long story short I try to put some contingencies into the case argument.  feel free to alter as you wish.
As for c span..... I hate c span.  the letter c sounds like alot of other letters.  the case arguments for c span could be and this is just a few:
c. span
C span
sea span
see span
si span
b span
be span
bee span
v span
e span

I removed "exit" this is an alexa native that ends the skill, the odds of this being passed on to lambda is nill  "end" "leave" do the same thing
I also removed Network from your custom slot values since "network is built into the utterances.
[customvalues.txt](https://github.com/bklavet/Echo-skill-to-control-Directv/files/1691408/customvalues.txt)
[index.txt](https://github.com/bklavet/Echo-skill-to-control-Directv/files/1691414/index.txt)

well see if this doesn't improve things.

**Comment by bklavet** [comment]
ok, double checked and network isn't built into the utterances, but it would make sense to add "network" to the utterances than the slot values.

that said, i think what I sent should work.  guess we will see.

**Comment by Ernster93** [comment]
Thank You... when I get a chance, I will change everything to the update and see how it all works.

**Comment by bklavet** [comment]
sure thing,  take a look at your original index and the changes i made,  The best way to test a channel that does not seem to be working is to test the name in the simulator.  Keep everything lower case, and see what is sent to lambda, and adjust accordingly.

Numbers usually sent with no problem so if a network name is not working and no matter what you cant get the station to work, you can tell it the number and that should be successful.  Or come up with another name.  example exchange NHL to hockey.

One final thought on locals channels.  get familiar with this in the browser
[localip]:8080/tv/getTuned (for main genie or standalone receiver)
[localip]:8080/tv/getTuned&clientAddr[MAC of Client](for mini ginei/client)

turn the receiver on to the channel you want and run the above command and you will get something like this for channel 11
![image](https://user-images.githubusercontent.com/9345832/35768770-077f0cc2-08c7-11e8-8d9f-3cdc553b03f8.png)

so use 11 instead of 011, or 0011.
